src/schemas.jl

```diff
@@ -0,0 +1,662 @@
+#####
+##### schema name/identifier parsing/validation
+#####
+
+const ALLOWED_SCHEMA_NAME_CHARACTERS = Char['-', '.', 'a':'z'..., '0':'9'...]
+
+"""
+    Legolas.is_valid_schema_name(x::AbstractString)
+
+Return `true` if `x` is a valid schema name, return `false` otherwise.
+
+Valid schema names are lowercase, alphanumeric, and may contain hyphens or periods.
+"""
+is_valid_schema_name(x::AbstractString) = all(i -> i in ALLOWED_SCHEMA_NAME_CHARACTERS, x)
+
+#####
+##### `SchemaVersion`
+#####
+
+"""
+    Legolas.SchemaVersion{name,version}
+
+A type representing a particular version of Legolas schema. The relevant `name` (a `Symbol`)
+and `version` (an `Integer`) are surfaced as type parameters, allowing them to be utilized for
+dispatch.
+
+For more details and examples, please see `Legolas.jl/examples/tour.jl` and the
+"Schema-Related Concepts/Conventions" section of the Legolas.jl documentation.
+
+The constructor `SchemaVersion{name,version}()` will throw an `ArgumentError` if `version` is
+not non-negative.
+
+See also: [`Legolas.@schema`](@ref)
+"""
+struct SchemaVersion{n,v}
+    function SchemaVersion{n,v}() where {n,v}
+        v isa Integer && v >= 0 || throw(ArgumentError("`version` in `SchemaVersion{_,version}` must be a non-negative integer, received: $v"))
```

Comment:
maybe this is helpful? In case there's a non-`Integer` that prints like an integer:
```suggestion
        v isa Integer && v >= 0 || throw(ArgumentError("`version` in `SchemaVersion{_,version}` must be a non-negative integer, received: $v (of type $(typeof(v))"))
```